### PR TITLE
Added listeners

### LINKED ISSUE #4: Add listeners to manage other NApp tables
This NApp needs these to be subscribe to these events:
- `kytos/flow_manager.flow.added`
- `kytos/of_core.handshake.completed`
- `kytos/flow_manager.flow.error`
- `kytos/[mef_eline|telemetry_int|coloring|of_lldp].enable_table`

Additionally, add `kytos/of_multi_table.enable_table` event to be published.

### PR DESCRIPTION
**Pull request by Alopalao**
### Summary

Added listeners and publication of the event `enable_table`
Process of how the operation of duplication flows starts has been implemented

### Local Tests

Enabled a pipeline to watch the events being sent.
Sent events manually
Currently NApps that support `enable_table` event:
- `coloring` PR [#44](https://github.com/kytos-ng/coloring/pull/44)
- `of_lldp` PR [#79](https://github.com/kytos-ng/of_lldp/pull/79)
- `mef_eline` PR [#321](https://github.com/kytos-ng/mef_eline/pull/321)

### End-to-End Tests

N/A

Dropping this Draft pull request since I created another one covering more than just events.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -120,6 +190,40 @@ def disable_pipeline(self, pipeline_id):
         log.debug("disable_pipeline result %s %s", msg, 201)
         return jsonify(msg), 201
 
+    def get_flows_to_be_installed(self):
```
Comment (Alopalao) [inline main.py:198]: Installing flows will start here but will be implemented in another PR.

File: main.py [hunk 2]
```diff
@@ -24,19 +25,64 @@ class Main(KytosNApp):
...
 
     def execute(self):
         """Execute once when the napp is running."""
         log.info("Executing of_multi_table")
 
+    def get_enabled_table(self):
+        """Get the only enabled table, if exists"""
+        return self.pipeline_controller.get_pipelines("enabled")
+
+    def load_enable_table(self, pipeline):
```
Comment (viniarck) [inline main.py]: I'm going to recommend that you start using typing hints on new added code, for existing code I typically don't ask, but for newer code we should be adding type hints to facilitate redability too, even though we're not linting that part yet.

File: main.py [hunk 3]
```diff
@@ -24,19 +25,64 @@ class Main(KytosNApp):
...
+        self.required_napps = found_napps
+        self.current_enabled_table = pipeline["pipelines"][0]
+        self.start_enabling_pipeline()
+
+    def get_enabled_napps(self):
+        """Get the NApps that are enabled and subscribed"""
+        enable_napps = set()
+        for key in self.controller.napps:
+            # Keys look like this: ('kytos', 'of_lldp')
+            if key[1] in SUBSCRIBED_NAPPS:
```
Comment (viniarck) [inline main.py]: It'd be handy to instead introduce `self.subscribed_napps` initializing it with `SUBSCRIBED_NAPPS` for testability, easier to mock.

File: main.py [hunk 4]
```diff
@@ -24,19 +25,64 @@ class Main(KytosNApp):
...
+                enable_napps.add(key[1])
+        return enable_napps
+
+    def start_enabling_pipeline(self):
+        """Method to start the process to enable table
+        First, build content to be sent through an event
+        """
+        content = {}
+        for table in self.current_enabled_table["multi_table"]:
```
Comment (viniarck) [inline main.py]: Depending how `start_enabling_pipeline` and `load_enable_table` are called you'd end up with a Thread race conditions since you're sharing `current_enabled_table`, you might need a flag or status with a threading Lock, we probably don't want this being potentially be executed from multiple threads once `POST /api/kytos/of_multi_table/v1/pipeline/<pipeline_id>/enable` could also enable a pipeline.

**Review by viniarck — COMMENTED** [review]
Left some initial feedback, I'll do another pass once the code is more structured.

File: main.py [hunk 5]
```diff
@@ -103,6 +150,28 @@ def enable_pipeline(self, pipeline_id):
...
+    @listen_to("kytos/(mef_eline|coloring|of_lldp).enable_table")
+    def on_enable_table(self, event):
+        """Listen for NApps responses"""
+        self.handle_enable_table(event)
+        
+    def handle_enable_table(self, event):
+        """Handle NApps responses from enable_table"""
+        self.event = event
+        napp = event.name.split('/')[1].split('.')[0]
+        self.required_napps.remove(napp)
```
Comment (Alopalao) [inline main.py:178]: I have a concern here. Do I check what is in the content and check what I have inside `self.current_enabled_table`? For example:
- `mef_eline` is running late with the response of a first attempt to enable a pipeline.
- The user enables a second pipeline (resetting the values from `self.required_napps`) with different set up for `mef_eline`.

Is it possible that `of_multi_table` receives the NApps' responses from the second enabling plus the response from `mef_eline` first enabling? I could have another variable `self.current_table_groups` with the last table groups to check content of responses.

Comment (viniarck) [reply main.py:178]: > The user enables a second pipeline (resetting the values from self.required_napps) with different set up for mef_eline.

It can happen, but we for the expected use case the application shouldn't allow it.

When enabling a pipeline, either you use a NApp instance level lock not to allow other ones to try to be enabled, or you can leverage the `status` field of each stored pipeline, and whenever you start to enabling a pipeline then you set an intermediary status accordingly such as `"enabling"` (instead of only "enabled" and "disabled"), since MongoDB ensures a lock then before trying to enable any NApp you query if there's any enabling NApp, if there is then you raise HTTP 409 conflict error explaining to the user that a pipeline X is still being enabled and only one can be enabled at a time.

> mef_eline is running late with the response of a first attempt to enable a pipeline.

Yes, it can happen. Eventually it's expected for `mef_eline` to reply though.

Comment (viniarck) [reply main.py:178]: Regarding `self.required_napps`, `self.current_enabled_table` and `self.current_table_groups`. Notice that you're modelling them as instance level global without being aware of pipeline, since we'll allow only a single pipeline to be enabled then it wouldn't cause many issues to debug. I'll encourage you to take a step back, and think if they all are even needed in the first place, based on [how to enable a pipeline](https://github.com/kytos-ng/kytos/blob/master/docs/blueprints/EP036.rst#enabling-pipeline), `self.current_enabled_table` and `self.current_table_groups` don't seem necessary, between steps 3 and 4, when disabling the current (soon to be previous) pipeline then you can get it and then store it locally and then once step 4 is completed you can compute the algorithm you'll implement to see which flows need to be installed or removed:

3. Disable the previous enabled pipeline, if there was any.
4. of_multi_table requests all flows from flow_manager API.

Comment (viniarck) [reply main.py:178]: Whenever applicable try to rely on the DB while avoiding to create new shared instance-level data instructures, unless absolutely needed. The reason why mef_eline and other NApps use to "load" a lot of things in memory were because back then it could rely on a full-fledged DB like Mongo.

File: main.py (hunk 4 above)
Comment (Alopalao) [reply main.py]: Got it, added

File: main.py (hunk 3 above)
Comment (Alopalao) [reply main.py]: Changed

File: main.py (hunk 2 above)
Comment (Alopalao) [reply main.py]: Ok, added.

File: main.py (hunk 5 above)
Comment (Alopalao) [reply main.py:178]: These change were made into another PR.

**Comment by Alopalao** [comment]
Made a larger PR addressing pipeline installation, #13 .

**Comment by viniarck** [comment]
> Made a larger PR addressing pipeline installation, #13 .

Thanks for the heads up @Alopalao, I've just reviewed it

**Comment by viniarck** [comment]
Closing this PR in favor of https://github.com/kytos-ng/of_multi_table/pull/13
